### ISSUE TITLE
Add support for DR Priority option in PIM hello messages

### ISSUE THREAD

**Issue by sesse** [troglobit/pimd#5]
When an IOS router sends out PIM hellos, they include a variety of options:

01:50:17.956481 IP (tos 0xc0, ttl 1, id 3442, offset 0, flags [none], proto PIM (103), length 58) 129.241.93.9 > 224.0.0.13: PIMv2, length 38
	Hello, cksum 0x7333 (correct)
	  Hold Time Option (1), length 2, Value: 1m45s
	    0x0000:  0069
	  Generation ID Option (20), length 4, Value: 0x9b4bd1df
	    0x0000:  9b4b d1df
	  DR Priority Option (19), length 4, Value: 0
	    0x0000:  0000 0000
	  State Refresh Capability Option (21), length 4, Value: v1
	    0x0000:  0100 0000
	  Unknown Option (65004), length 0, Value: 

In particular, the DR Priority Option helps other routers determine which one is the DR for this subnet. However, pimd does not include this option, or barely any at all:

01:50:17.961914 IP (tos 0x0, ttl 1, id 33270, offset 0, flags [none], proto PIM (103), length 30) 129.241.93.30 > 224.0.0.13: PIMv2, length 10
	Hello, cksum 0xdf93 (correct)
	  Hold Time Option (1), length 2, Value: 1m45s
	    0x0000:  0069

This makes it seemingly always win the election, even when it shouldn't. The included example pimd.conf claims setting preference to 101 should make it lose to other gear, but in practice it seems to have no effect.

**Comment by troglobit** [comment]
Support now (finally) in progress for this.  Will likely be part of the upcoming v2.3.0 release.

**Comment by idismmxiv** [comment]
Great work @troglobit, small nitpick though:

If I interpret RFC 4601 correctly, then your implementation makes false assumption that DR Priority 0 would be the same as missing DR Priority field all together. I believe that if there is a DR Priority with value 0 it is no more special than value 1, 5 or any other and should be treated just as the lowest priority possible. So the check in the code that if neighbor (nbr->dr_prio == 0) should be replaced with check (!nbr->dr_prio_present).

So in short:
(nbr->dr_prio = 0) != (!nbr->dr_prio)

No patch at this moment sorry, but I could come up with one if you agree with my interpretation or you can make the change yourself. :)

**Comment by troglobit** [comment]
Interesting, @idismmxiv ... that I must look into!  I've had some problems during testing of my last
features/fixes on the dev branch, what you mention may be the root cause.  It also gives me a reason
to refactor the `parse_pim_hello()` function! :smiley:

**Comment by troglobit** [comment]
There, refactored ... but I still have some regression in my test setup. Can't seem to get the RP to forward multicast to a neighbor router atm. :frowning:

**Comment by troglobit** [comment]
Nope, turns out it was a problem with my virtual test bed! :innocent: :smiley: 

Support for dr-priority has now been merged to master, soon to be released as pimd v2.3.0!
